File: cache-validate-mapper/src/main/java/com/example/carlosjr/cachevalidatemapper/exceptions/GlobalExceptionHandler.java
package com.example.carlosjr.cachevalidatemapper.exceptions;

import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.server.MethodNotAllowedException;

import java.time.LocalDateTime;

@ControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ExceptionEnhanceDto> genericExceptionHandler(Exception ex, HttpServletRequest request){

        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(getEnhancedResponse(
                        ex.getMessage(),
                        request.getRequestURI(),
                        ex.getClass().getName()));

    }

    @ExceptionHandler(TeamNotFoundException.class)
    public ResponseEntity<ExceptionEnhanceDto> handleTeamNotFound(TeamNotFoundException ex, HttpServletRequest request){

        return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(getEnhancedResponse(
                        ex.getMessage(),
                        request.getRequestURI(),
                        ex.getClass().getName()));
    }

    @ExceptionHandler(MethodNotAllowedException.class)
    public ResponseEntity<ExceptionEnhanceDto> handleMethodNotAllowed(MethodNotAllowedException ex, HttpServletRequest request){

        return ResponseEntity.status(HttpStatus.METHOD_NOT_ALLOWED)
                .body(getEnhancedResponse(
                        ex.getMessage(),
                        request.getRequestURI(),
                        ex.getClass().getName()));
    }

    private ExceptionEnhanceDto getEnhancedResponse(String msg, String path, String className){
        return ExceptionEnhanceDto.builder()
                .message(msg)
                .path(path)
                .className(className)
                .timestamp(LocalDateTime.now())
                .build();
    }



}

File: basic-mvc-kotlin/src/main/kotlin/com/example/carlos/basicmvckotlin/exceptions/GlobalExceptionHandler.kt
package com.example.carlos.basicmvckotlin.exceptions

import jakarta.servlet.http.HttpServletRequest
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import java.time.LocalDateTime

@ControllerAdvice
class GlobalExceptionHandler {
    @ExceptionHandler(ResourceNotFoundException::class)
    fun handleResourceNotFound(exception: ResourceNotFoundException,
                               httpRequest: HttpServletRequest) : ResponseEntity<ExceptionDto> {

        val exceptionDto = getExceptionDto(
                exception.message,
                httpRequest.requestURI,
                HttpStatus.NOT_FOUND.value())

        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(exceptionDto)

    }

    @ExceptionHandler(DataIntegrityViolationException::class)
    fun handleDataIntegrityViolation(exception: DataIntegrityViolationException,
                                     httpRequest: HttpServletRequest) : ResponseEntity<ExceptionDto>{

        val exceptionDto = getExceptionDto(
                exception.message,
                httpRequest.requestURI,
                HttpStatus.BAD_REQUEST.value())

        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exceptionDto)
    }

    fun getExceptionDto(msg: String?, path: String?, code: Int?) : ExceptionDto{
        val exceptionDto = ExceptionDto()
        exceptionDto.message = msg
        exceptionDto.path = path
        exceptionDto.responseCode = code
        exceptionDto.timestamp = LocalDateTime.now()

        return exceptionDto
    }
}